File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/sdk/screen/RewardedMethodHandler.kt
package com.cleveradssolutions.plugin.flutter.sdk.screen

import com.cleveradssolutions.plugin.flutter.CASFlutterContext
import com.cleveradssolutions.plugin.flutter.bridge.base.AdMethodHandler
import com.cleveradssolutions.plugin.flutter.sdk.AdContentInfoMethodHandler
import com.cleveradssolutions.plugin.flutter.sdk.OnAdImpressionListenerHandler
import com.cleveradssolutions.plugin.flutter.util.getArgAndReturn
import com.cleveradssolutions.plugin.flutter.util.success
import com.cleveradssolutions.sdk.screen.CASRewarded
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

private const val CHANNEL_NAME = "cleveradssolutions/rewarded"

class RewardedMethodHandler(
    binding: FlutterPlugin.FlutterPluginBinding,
    private val contextService: CASFlutterContext,
    contentInfoHandler: AdContentInfoMethodHandler
) : AdMethodHandler<CASRewarded>(binding, CHANNEL_NAME, contentInfoHandler) {

    override fun initInstance(id: String): Ad<CASRewarded> {
        val context = contextService.getContext()
        val rewarded = CASRewarded(context, id)
        val contentInfoId = "rewarded_$id"
        rewarded.contentCallback = ScreenAdContentCallbackHandler(this, id, contentInfoId)
        rewarded.onImpressionListener = OnAdImpressionListenerHandler(this, id, contentInfoId)
        return Ad(rewarded, id, contentInfoId)
    }

    override fun onMethodCall(
        instance: Ad<CASRewarded>,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        when (call.method) {
            "isAutoloadEnabled" -> isAutoloadEnabled(instance.ad, result)
            "setAutoloadEnabled" -> setAutoloadEnabled(instance.ad, call, result)
            "isExtraFillInterstitialAdEnabled" ->
                isExtraFillInterstitialAdEnabled(instance.ad, result)

            "setExtraFillInterstitialAdEnabled" ->
                setExtraFillInterstitialAdEnabled(instance.ad, call, result)

            "isLoaded" -> isLoaded(instance.ad, result)
            "getContentInfo" -> getContentInfo(instance.contentInfoId, result)
            "load" -> load(instance.ad, result)
            "show" -> show(instance, result)
            "destroy" -> destroy(instance, result)
            else -> super.onMethodCall(instance, call, result)
        }
    }

    private fun isAutoloadEnabled(rewarded: CASRewarded, result: MethodChannel.Result) {
        result.success(rewarded.isAutoloadEnabled)
    }

    private fun setAutoloadEnabled(
        rewarded: CASRewarded,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        call.getArgAndReturn<Boolean>("isEnabled", result) {
            rewarded.isAutoloadEnabled = it
        }
    }

    private fun isExtraFillInterstitialAdEnabled(
        rewarded: CASRewarded,
        result: MethodChannel.Result
    ) {
        result.success(rewarded.isExtraFillInterstitialAdEnabled)
    }

    private fun setExtraFillInterstitialAdEnabled(
        rewarded: CASRewarded,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        call.getArgAndReturn<Boolean>("isEnabled", result) {
            rewarded.isExtraFillInterstitialAdEnabled = it
        }
    }

    private fun isLoaded(rewarded: CASRewarded, result: MethodChannel.Result) {
        result.success(rewarded.isLoaded)
    }

    private fun getContentInfo(contentInfoId: String, result: MethodChannel.Result) {
        result.success(contentInfoId)
    }

    private fun load(rewarded: CASRewarded, result: MethodChannel.Result) {
        rewarded.load(contextService.getContext())

        result.success()
    }

    private fun show(ad: Ad<CASRewarded>, result: MethodChannel.Result) {
        val activity = contextService.getActivity()

        val onRewardEarnedListener = OnRewardEarnedListenerHandler(this, ad.id, ad.contentInfoId)
        ad.ad.show(activity, onRewardEarnedListener)

        result.success()
    }

    private fun destroy(ad: Ad<CASRewarded>, result: MethodChannel.Result) {
        destroy(ad)
        ad.ad.destroy()

        result.success()
    }

}

File: ios/Classes/SDK/Screen/RewardedMethodHandler.swift
//
//  RewardedMethodHandler.swift
//  clever_ads_solutions
//
//  Copyright © 2025 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import CleverAdsSolutions
import Flutter

private let channelName = "cleveradssolutions/rewarded"

class RewardedMethodHandler: AdMethodHandler<CASRewarded> {
    private var contentDelegate: ScreenContentDelegateHandler?
    private var impressionDelegate: ImpressionDelegateHandler?
    
    init(with registrar: FlutterPluginRegistrar, _ contentInfoMethodHandler: AdContentInfoMethodHandler) {
        super.init(with: registrar, on: channelName, contentInfoMethodHandler)
    }

    override func initInstance(_ id: String) -> AdMethodHandler<CASRewarded>.Ad {
        let rewarded = CASRewarded(casID: id)
        let contentInfoId = "rewarded_\(id)"
        let contentDelegate = ScreenContentDelegateHandler(self, id, contentInfoId)
        self.contentDelegate = contentDelegate
        rewarded.delegate = contentDelegate
        let impressionDelegate = ImpressionDelegateHandler(self, id, contentInfoId)
        rewarded.impressionDelegate = impressionDelegate
        self.impressionDelegate = impressionDelegate
        return Ad(ad: rewarded, id: id, contentInfoId: contentInfoId)
    }

    override func onMethodCall(_ instance: Ad, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        switch call.method {
        case "isAutoloadEnabled":
            isAutoloadEnabled(instance.ad, result)
        case "setAutoloadEnabled":
            setAutoloadEnabled(instance.ad, call, result)
        case "isExtraFillInterstitialAdEnabled":
            isExtraFillInterstitialAdEnabled(instance.ad, result)
        case "setExtraFillInterstitialAdEnabled":
            setExtraFillInterstitialAdEnabled(instance.ad, call, result)
        case "isLoaded":
            isLoaded(instance.ad, result)
        case "getContentInfo":
            getContentInfo(instance.contentInfoId, result)
        case "load":
            load(instance.ad, result)
        case "show":
            show(instance, result)
        case "destroy":
            destroy(instance, result)
        default: super.onMethodCall(call, result)
        }
    }

    private func isAutoloadEnabled(_ rewarded: CASRewarded, _ result: @escaping FlutterResult) {
        result(rewarded.isAutoloadEnabled)
    }

    private func setAutoloadEnabled(_ rewarded: CASRewarded, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("isEnabled", result) { isEnabled in
            rewarded.isAutoloadEnabled = isEnabled
        }
    }

    private func isExtraFillInterstitialAdEnabled(_ rewarded: CASRewarded, _ result: @escaping FlutterResult) {
//        result(rewarded.isExtraFillInterstitialAdEnabled)

        result(nil)
    }

    private func setExtraFillInterstitialAdEnabled(_ rewarded: CASRewarded, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
//        call.getArgAndReturn("isEnabled", result) { isEnabled in
//            rewarded.isExtraFillInterstitialAdEnabled = isEnabled
//        }

        result(nil)
    }

    private func isLoaded(_ rewarded: CASRewarded, _ result: @escaping FlutterResult) {
        result(rewarded.isAdLoaded)
    }

    private func getContentInfo(_ contentInfoId: String, _ result: @escaping FlutterResult) {
        result(contentInfoId)
    }

    private func load(_ rewarded: CASRewarded, _ result: @escaping FlutterResult) {
        rewarded.loadAd()

        result(nil)
    }

    private func show(_ ad: AdMethodHandler<CASRewarded>.Ad, _ result: @escaping FlutterResult) {
        let userDidEarnRewardHandler = UserDidEarnRewardDelegateHandler(self, ad.id, ad.contentInfoId)
        let id = ad.id
        ad.ad.present(userDidEarnRewardHandler: { info in
            invokeMethod("onUserEarnedReward", ["contentInfoId": contentInfoId])
        })

        result(nil)
    }

    private func destroy(_ ad: AdMethodHandler<CASRewarded>.Ad, _ result: @escaping FlutterResult) {
        contentDelegate = nil
        impressionDelegate = nil
        destroy(ad: ad)
        ad.ad.destroy()

        result(nil)
    }
}
